src/plots/histogram.rs

```diff
@@ -0,0 +1,440 @@
+use crate::error;
+use crate::plots::{Plot, PlotData};
+use crate::primitives::{DataRef, FeatureDataRef, Measurement, NullableDataRef};
+use crate::util::Result;
+use float_cmp::*;
+use ndarray::{stack, Array, Array1, Axis};
+use serde::{Deserialize, Serialize};
+use snafu::ensure;
+use std::collections::HashMap;
+use std::iter::FromIterator;
+use vega_lite_3::{
+    BinEnum, EncodingBuilder, Mark, Padding, SelectionDefBuilder, SelectionDefType,
+    SingleDefUnitChannel, StandardType, VegaliteBuilder, X2ClassBuilder, XClassBuilder,
+    YClassBuilder,
+};
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Histogram {
+    counts: Vec<u64>,
+    labels: Option<Vec<String>>,
+    nodata_count: u64,
+    min: f64,
+    max: f64,
+    measurement: Measurement,
+}
+
+impl Histogram {
+    /// Creates a new empty histogram
+    ///
+    /// # Examples
+    /// ```rust
+    /// use geoengine_datatypes::plots::Histogram;
+    /// use geoengine_datatypes::primitives::Measurement;
+    /// use std::f64;
+    ///
+    /// Histogram::new(2, 0., 1., Measurement::Unitless).unwrap();
+    ///
+    /// Histogram::new(0, f64::NAN, f64::INFINITY, Measurement::Unitless).unwrap_err();
+    /// ```
+    pub fn new(
+        number_of_buckets: usize,
+        min: f64,
+        max: f64,
+        measurement: Measurement,
+    ) -> Result<Self> {
+        ensure!(
+            number_of_buckets > 0,
+            error::PlotError {
+                details: "Histograms must have at least one bucket"
+            }
+        );
+        ensure!(
+            min.is_finite() && max.is_finite(),
+            error::PlotError {
+                details: "Histograms must have finite min/max values"
+            }
+        );
+        ensure!(
+            min < max || (approx_eq!(f64, min, max) && number_of_buckets == 1),
+            error::PlotError {
+                details: "Histograms max value must be larger than its min value"
+            }
+        );
+
+        Ok(Self {
+            counts: vec![0; number_of_buckets],
+            labels: None,
+            nodata_count: 0,
+            min,
+            max,
+            measurement,
+        })
+    }
+
+    /// Creates a new empty histogram
+    ///
+    /// # Examples
+    /// ```rust
+    /// use geoengine_datatypes::plots::Histogram;
+    /// use geoengine_datatypes::primitives::Measurement;
+    /// use std::f64;
+    ///
+    /// Histogram::new_with_labels(2, 0., 1., Measurement::Unitless, vec!["foo".into(), "bar".into()]).unwrap();
+    ///
+    /// Histogram::new_with_labels(0, f64::NAN, f64::INFINITY, Measurement::Unitless, vec!["foo".into(), "bar".into()]).unwrap_err();
+    /// Histogram::new_with_labels(2, 0., 1., Measurement::Unitless, vec!["foo".into()]).unwrap_err();
+    /// ```
+    pub fn new_with_labels(
```

Comment:
crys for builder pattern

Reply:
be31753

Reply:
goes in the right direction, but a builder should have a function for every value. Maybe we could use something to automate this like: https://github.com/colin-kiegel/rust-derive-builder

Reply:
We talked about this since the parameters do not have good default values. We could, in a separate task, use a builder framework for all builders. This is not the only one. But we first need our own definition of how to build builders.